exp/kubeadm-ignition/internal/ignition/s3backend_backend.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ignition
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"github.com/aws/aws-sdk-go/aws"
```

Comment:
https://github.com/kubernetes-sigs/cluster-api-provider-awswhy are there AWS specific things in here? Any AWS provider code should go in CAPA https://github.com/kubernetes-sigs/cluster-api-provider-aws

Reply:
From what I understand, Ignition needs a source from which it pulls its configuration. This implementation is using S3 at the moment. We may want/need to support multiple sources, and S3 could/would be a reasonable choice when running on AWS. I don't know there's any simple way to keep source-specific code out of the bootstrap provider, unless you went with a plugin model, where the plugins lived in separate repos. That, however, would make deployment a lot more complicated.

Reply:
A bootstrap provider that lives in the Cluster API repo being infrastructure agnostic should be a requirement IMO otherwise we're setting a precedent and breaking one of the core principles of Cluster API. If it's not straightforward it should come with a design proposal.

"I don't know there's any simple way to keep source-specific code out of the bootstrap provider"  --> sounds like we need to do a bit more research around this.

Reply:
https://github.com/flatcar-linux/ignition/blob/master/doc/supported-platforms.md gives a good overview of using ignition with different platforms

Reply:
+1 making this agnostic, let's discuss and design a solution that is agnostic. We have prior art with object references, or we could make this based on the infrastructure provider used (infra providers would provide the code that runs here)

Reply:
Possibly 🤔  We try to be as agnostic as possible, what that really comes down to is provider interfaces that are satisfied with third party installations.

For example, we could abstract away this interface to be a `metadata provider`. Then, the AWS provider could contribute the AWS code, GCP, Azure can do the same with time. 

What do you all think?

Reply:
this might should be part of the https://github.com/kubernetes-sigs/cluster-api/issues/3761 design